solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -148,29 +127,188 @@ library Groups {
+    ///         active and checks if it hasn't expired. If so, updates the information
+    ///         about expired groups so that all expired groups are marked as such.
+    function expireOldGroups(Data storage self) internal {
+        // Move expiredGroupOffset as long as there are some groups that should
+        // be marked as expired. It is possible that expired group offset will
+        // move out of the groups array by one position. It means that all groups
+        // are expired (it points to the first active group) and that place in
+        // groups array - currently empty - will be possibly filled later by
+        // a new group.
+        while (
+            self.expiredGroupOffset < self.groupsRegistry.length &&
+            groupLifetimeOf(
+                self,
+                self.groupsRegistry[self.expiredGroupOffset]
+            ) <
+            block.number
+        ) {
+            self.expiredGroupOffset++;
+            self.activeGroupsCount--;
+        }
 
-    /// @notice Gets the number of active groups. Candidate, expired and terminated
-    ///         groups are not counted as active.
-    function numberOfActiveGroups(Data storage self)
+        // Go through all activeTerminatedGroups and if some of the terminated
+        // groups are expired, remove them from activeTerminatedGroups collection
+        // and rearrange the array to preserve the original order.
+        // This is needed because we evaluate the shift of selected group index
+        // based on how many non-expired groups have been terminated. Hence it is
+        // important that a number of terminated groups matches the length of
+        // activeTerminatedGroups[].
+        for (uint256 i = 0; i < self.activeTerminatedGroups.length; i++) {
+            if (self.expiredGroupOffset > self.activeTerminatedGroups[i]) {
+                // When 'i'th group qualifies for expiration, we need to remove
+                // it from the activeTerminatedGroups array manually by rearranging
+                // the order starting from 'i'th group.
+                for (
+                    uint256 j = i;
+                    j < self.activeTerminatedGroups.length - 1;
+                    j++
+                ) {
+                    self.activeTerminatedGroups[j] = self
+                        .activeTerminatedGroups[j + 1];
+                }
+                // Resizing the array length by 1. The last element was copied
+                // over in the loop above to an index "second to last". This is
+                // why we can safely remove it from here.
+                self.activeTerminatedGroups.pop();
+                // At this point the length of activeTerminatedGroups[] was shrinked
+                // by 1. We need to adjust 'i'th counter by 1 as well, otherwise
+                // it will overflow this array and throw an error.
+                i--;
```

Comment:
What if `i=0`? Say we have `activeTerminatedGroups = [10, 15, 50, 90]` and `self.expiredGroupOffset = 11`.

Reply:
I realized I didn't move the tests from https://github.com/keep-network/keep-core-ghsa-m4mj-gfw9-8gj5/pull/2/files After I'm done with that, I'll have an answer if that's a bug or not.

Reply:
It was a bug. The tests didn't catch it in v1, because the contracts were written in Solidity 0.5.17 which "allowed" going below 0. I checked it on Remix and `0-1` resulted in uin256 max number. Whereas in Solidity 0.8.6 it reverts. Tests that were moved from the security repo branch also confirmed that on my local. Fix + tests are here b57f948